File: services/order/internal/tracing/tracing.go
// Package tracing initializes OpenTelemetry distributed tracing for the
// Order service, per the "Phase 4: distributed tracing" section of
// shared/CONTRACTS.md.
//
// Conventions implemented here (all from CONTRACTS.md, not invented):
//   - Exporter protocol: OTLP over HTTP, reading the endpoint from the
//     standard OTEL_EXPORTER_OTLP_ENDPOINT env var (default
//     http://localhost:4318 for local dev if unset).
//   - Service name: standard OTEL_SERVICE_NAME env var (default
//     "order-service" if unset).
//   - Sampling: always-on (100%) — a deliberate simplification for this
//     low-traffic practice app, not a production default.
//   - Propagation format: W3C Trace Context, via the SDK's built-in
//     propagator (propagation.TraceContext) — no custom scheme.
//
// Both OTEL_SERVICE_NAME and OTEL_EXPORTER_OTLP_ENDPOINT are standard,
// spec-defined env vars that the SDK already knows how to read on its own
// (resource.WithFromEnv() reads OTEL_SERVICE_NAME; otlptracehttp.New reads
// OTEL_EXPORTER_OTLP_ENDPOINT / OTEL_EXPORTER_OTLP_TRACES_ENDPOINT). This
// package only seeds process defaults for the two vars when unset so the
// SDK's own env-reading logic picks up this service's documented defaults
// without needing any explicit WithEndpoint/WithServiceName option calls.
package tracing

import (
	"context"
	"fmt"
	"os"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
	"go.opentelemetry.io/otel/propagation"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	"go.uber.org/zap"
)

const (
	defaultServiceName  = "order-service"
	defaultOTLPEndpoint = "http://localhost:4318"

	otelServiceNameEnv  = "OTEL_SERVICE_NAME"
	otelOTLPEndpointEnv = "OTEL_EXPORTER_OTLP_ENDPOINT"
)

// Shutdown flushes and stops the tracer provider. Callers should invoke it
// during graceful shutdown, with a bounded-timeout context.
type Shutdown func(ctx context.Context) error

// Init sets up the global OTel TracerProvider and W3C Trace Context
// propagator for the process. It is additive instrumentation: per
// CONTRACTS.md, if the OTLP endpoint is unreachable, spans are dropped and
// export errors are logged as warnings — this must never block startup or
// crash the service, so Init itself only fails on genuine local
// misconfiguration (e.g. a malformed endpoint URL), not on the collector
// being down (export happens asynchronously, after Init returns).
func Init(ctx context.Context, logger *zap.Logger) (Shutdown, error) {
	// Seed documented defaults for the two standard env vars this service
	// reads, per CONTRACTS.md, before the SDK's own env detectors run.
	if os.Getenv(otelServiceNameEnv) == "" {
		_ = os.Setenv(otelServiceNameEnv, defaultServiceName)
	}
	if os.Getenv(otelOTLPEndpointEnv) == "" {
		_ = os.Setenv(otelOTLPEndpointEnv, defaultOTLPEndpoint)
	}

	// Route the SDK's internal error handler (used for async export
	// failures, e.g. an unreachable collector) through our structured
	// logger instead of the default stderr writer, at warn level per
	// CONTRACTS.md ("logged as a warning, never block or crash").
	otel.SetErrorHandler(otel.ErrorHandlerFunc(func(err error) {
		logger.Warn("otel export error", zap.Error(err))
	}))

	res, err := resource.New(ctx,
		resource.WithFromEnv(),      // OTEL_SERVICE_NAME / OTEL_RESOURCE_ATTRIBUTES
		resource.WithTelemetrySDK(), // telemetry.sdk.{name,language,version}
		resource.WithHost(),         // host.name
	)
	if err != nil {
		return nil, fmt.Errorf("build otel resource: %w", err)
	}

	// otlptracehttp.New reads OTEL_EXPORTER_OTLP_ENDPOINT (and
	// OTEL_EXPORTER_OTLP_TRACES_ENDPOINT, which takes precedence) itself;
	// no WithEndpoint option is passed here so the standard env var stays
	// authoritative, per CONTRACTS.md. An http:// (not https://) scheme in
	// the endpoint is auto-detected by the SDK as an insecure connection.
	exporter, err := otlptracehttp.New(ctx)
	if err != nil {
		return nil, fmt.Errorf("build otlp http exporter: %w", err)
	}

	tp := sdktrace.NewTracerProvider(
		sdktrace.WithBatcher(exporter),
		sdktrace.WithResource(res),
		sdktrace.WithSampler(sdktrace.AlwaysSample()),
	)

	otel.SetTracerProvider(tp)
	otel.SetTextMapPropagator(propagation.TraceContext{})

	logger.Info("otel tracing initialized",
		zap.String("serviceName", os.Getenv(otelServiceNameEnv)),
		zap.String("otlpEndpoint", os.Getenv(otelOTLPEndpointEnv)),
		zap.String("sampler", "always_on"),
	)

	return tp.Shutdown, nil
}

File: services/payment/app/tracing.py
"""OpenTelemetry SDK setup for the Payment service (Phase 4).

Per CONTRACTS.md's "Phase 4: distributed tracing" section:
  - Exporter protocol: OTLP over HTTP (`opentelemetry-exporter-otlp-proto-http`).
  - Endpoint: standard env var `OTEL_EXPORTER_OTLP_ENDPOINT`, default
    `http://localhost:4318`. The exporter class reads this env var itself when no
    `endpoint=` kwarg is passed (verified against the installed SDK: `OTLPSpanExporter()`
    falls back to `environ[OTEL_EXPORTER_OTLP_ENDPOINT]` or `http://localhost:4318/` and
    appends `/v1/traces`) -- we do NOT need to read/parse this var ourselves.
  - Service name: standard env var `OTEL_SERVICE_NAME`. `Resource.create()` also reads
    this itself, but its own built-in fallback is `unknown_service`, not our desired
    per-service default. So we `setdefault` the env var to `payment-service` *before*
    calling `Resource.create()`, which both (a) respects an operator-supplied
    `OTEL_SERVICE_NAME` and (b) gives us the right default when unset -- rather than
    passing an explicit `{"service.name": ...}` attribute, which would unconditionally
    override the env var even when one was supplied (confirmed via a quick throwaway
    script against opentelemetry-sdk 1.43.0: `Resource.create({SERVICE_NAME: "x"})`
    always wins over `OTEL_SERVICE_NAME`, so setdefault-then-create is the only way to
    get "env var wins if set, else our default").
  - Sampling: always-on (100%) -- this SDK's default `Sampler` (`ParentBased(ALWAYS_ON)`)
    already does this, no explicit sampler config needed. Documented as a deliberate
    practice-app simplification in README.md.
  - Propagation: W3C Trace Context, via the OTel API's default global propagator
    (`opentelemetry.propagate`) -- also already the SDK default when no
    `OTEL_PROPAGATORS` env var is set, nothing to configure here.
  - If the OTLP endpoint is unreachable, spans must be dropped/logged as a warning, never
    block or crash the service. This is inherent behavior of `BatchSpanProcessor` (runs
    export in a background thread) and `OTLPSpanExporter` (catches request exceptions in
    `export()` and returns `SpanExportResult.FAILURE` instead of raising) -- no extra
    handling required here.

Must be called as early as possible in the entrypoint (see main.py), before importing
anything that should be auto-instrumented (FastAPI), mirroring the existing
`configure_logging()` early-import pattern in main.py.
"""
from __future__ import annotations

import os

from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor

DEFAULT_SERVICE_NAME = "payment-service"

_configured = False


def configure_tracing() -> None:
    """Initialize the global TracerProvider with an OTLP/HTTP exporter.

    Idempotent -- safe to call more than once (e.g. under a test harness that
    re-imports the app), only the first call takes effect.
    """
    global _configured
    if _configured:
        return

    os.environ.setdefault("OTEL_SERVICE_NAME", DEFAULT_SERVICE_NAME)

    resource = Resource.create()
    provider = TracerProvider(resource=resource)

    # OTLPSpanExporter() with no endpoint kwarg reads OTEL_EXPORTER_OTLP_ENDPOINT
    # itself (default http://localhost:4318, per CONTRACTS.md) -- verified above.
    otlp_exporter = OTLPSpanExporter()
    provider.add_span_processor(BatchSpanProcessor(otlp_exporter))

    # Optional, off by default: a console exporter for local verification, per
    # CONTRACTS.md's "wire up (at least temporarily, for your own verification) a
    # console span exporter" instruction. Not part of the required contract env
    # vars -- purely a debugging aid so spans are visible without a running
    # collector. Enable with OTEL_CONSOLE_EXPORTER=true.
    if os.getenv("OTEL_CONSOLE_EXPORTER", "").strip().lower() in ("1", "true", "yes"):
        from opentelemetry.sdk.trace.export import ConsoleSpanExporter, SimpleSpanProcessor

        provider.add_span_processor(SimpleSpanProcessor(ConsoleSpanExporter()))

    trace.set_tracer_provider(provider)
    _configured = True


def get_tracer() -> trace.Tracer:
    return trace.get_tracer("payment-service")
